problems with make -j8

Sometimes when building pkgsrc/wip/unison-snapshot (which currently corresponds to git commit 08dbd399b65b960e7f2ea356555209404d95024a) I see random-looking errors. My best guess is that the Makefiles do not track all necessary dependencies and some jobs run in parallel that shouldn't.

At the time of the test, the machine was not CPU or disk limited AFAIK.

The build immediately before this one completed successfully, with all the same settings.

Here is one such failing build log:

[log.txt](https://github.com/bcpierce00/unison/files/9304261/log.txt)

It ends with:
```
Linking unison
ocamlopt -verbose -g -I lwt -I ubase -I system -I +unix -I +str -I system/generic -I lwt/generic  -o unison   unix.cmxa str.cmxa ubase/umarshal.cmx ubase/rx.cmx unicode_tables.cmx unicode.cmx bytearray.cmx system/system_generic.cmx system/generic/system_impl.cmx system.cmx ubase/projectInfo.cmx ubase/myMap.cmx ubase/safelist.cmx ubase/util.cmx ubase/uarg.cmx ubase/prefs.cmx ubase/trace.cmx ubase/proplist.cmx lwt/pqueue.cmx lwt/lwt.cmx lwt/lwt_util.cmx lwt/generic/lwt_unix_impl.cmx lwt/lwt_unix.cmx features.cmx uutil.cmx case.cmx pred.cmx fileutil.cmx name.cmx path.cmx fspath.cmx fs.cmx fingerprint.cmx abort.cmx osx.cmx external.cmx fswatch.cmx props.cmx fileinfo.cmx os.cmx lock.cmx clroot.cmx common.cmx tree.cmx checksum.cmx terminal.cmx transfer.cmx xferhint.cmx remote.cmx negotiate.cmx globals.cmx fswatchold.cmx fpcache.cmx update.cmx copy.cmx stasher.cmx files.cmx sortri.cmx recon.cmx transport.cmx strings.cmx uicommon.cmx uitext.cmx test.cmx main.cmx linktext.cmx osxsupport.o pty.o bytearray_stubs.o hash_compat.o -cclib -lutil
ocamlopt: uitext.ml ---> uitext.cmx
ocamlopt -g -I lwt -I ubase -I system -I +unix -I +str -I system/generic -I lwt/generic -I /usr/pkg/lib/ocaml/site-lib/lablgtk3 -I /usr/pkg/lib/ocaml/site-lib/cairo2 -c /scratch/wip/unison-snapshot/work/unison-08dbd399b65b960e7f2ea356555209404d95024a/src/uitext.ml
ocamlopt: test.ml ---> test.cmx
ocamlopt -g -I lwt -I ubase -I system -I +unix -I +str -I system/generic -I lwt/generic -I /usr/pkg/lib/ocaml/site-lib/lablgtk3 -I /usr/pkg/lib/ocaml/site-lib/cairo2 -c /scratch/wip/unison-snapshot/work/unison-08dbd399b65b960e7f2ea356555209404d95024a/src/test.ml
+ gcc -c  -o '/tmp/camlstartupfa1c77.o' '/tmp/camlstartup4cc39c.s'
+ gcc -O2 -fno-strict-aliasing -fwrapv  -pthread  -Wall -Wdeclaration-after-statement -fno-common -fexcess-precision=standard -fno-tree-vrp -fPIC -ffunction-sections -O2 -g -D_FORTIFY_SOURCE=2 -fstack-clash-protection -I/usr/include -Wl,-E -Wl,-zrelro -Wl,-znow -L/usr/lib -Wl,-R/usr/lib -Wl,-R/usr/pkg/lib -o 'unison'  '-Llwt' '-Lubase' '-Lsystem' '-L/usr/pkg/lib/ocaml/unix' '-L/usr/pkg/lib/ocaml/str' '-Lsystem/generic' '-Llwt/generic' '-L/usr/pkg/lib/ocaml' -O2 -g -D_FORTIFY_SOURCE=2 -fstack-clash-protection -I/usr/pkg/include -I/usr/include -I/usr/pkg/include/glib-2.0 -I/usr/pkg/include/gio-unix-2.0 -I/usr/pkg/lib/glib-2.0/include -I/usr/pkg/include/freetype2 -I/usr/pkg/include/harfbuzz -I/usr/pkg/include/libdrm -Wl,-zrelro -Wl,-znow -L/usr/pkg/lib -Wl,-R/usr/pkg/lib -L/usr/lib -Wl,-R/usr/lib '/tmp/camlstartupfa1c77.o' '/usr/pkg/lib/ocaml/std_exit.o' 'linktext.o' 'main.o' 'test.o' 'uitext.o' 'uicommon.o' 'strings.o' 'transport.o' 'recon.o' 'sortri.o' 'files.o' 'stasher.o' 'copy.o' 'update.o' 'fpcache.o' 'fswatchold.o' 'globals.o' 'negotiate.o' 'remote.o' 'xferhint.o' 'transfer.o' 'terminal.o' 'checksum.o' 'tree.o' 'common.o' 'clroot.o' 'lock.o' 'os.o' 'fileinfo.o' 'props.o' 'fswatch.o' 'external.o' 'osx.o' 'abort.o' 'fingerprint.o' 'fs.o' 'fspath.o' 'path.o' 'name.o' 'fileutil.o' 'pred.o' 'case.o' 'uutil.o' 'features.o' 'lwt/lwt_unix.o' 'lwt/generic/lwt_unix_impl.o' 'lwt/lwt_util.o' 'lwt/lwt.o' 'lwt/pqueue.o' 'ubase/proplist.o' 'ubase/trace.o' 'ubase/prefs.o' 'ubase/uarg.o' 'ubase/util.o' 'ubase/safelist.o' 'ubase/myMap.o' 'ubase/projectInfo.o' 'system.o' 'system/generic/system_impl.o' 'system/system_generic.o' 'bytearray.o' 'unicode.o' 'unicode_tables.o' '
ubase/rx.o' 'ubase/umarshal.o' '/usr/pkg/lib/ocaml/str.a' '/usr/pkg/lib/ocaml/unix.a' '/usr/pkg/lib/ocaml/stdlib.a' '-lcamlstr' '-lunix' 'osxsupport.o' 'pty.o' 'bytearray_stubs.o' 'hash_compat.o' '-lutil' '/usr/pkg/lib/ocaml/libasmrun.a' -lm
ld: /tmp/camlstartupfa1c77.o: in function `caml_startup__code_begin':
:(.text+0x675): undefined reference to `camlTest__entry'
ld: /tmp/camlstartupfa1c77.o: in function `caml_globals':
:(.data+0x590): undefined reference to `camlTest__gc_roots'
ld: /tmp/camlstartupfa1c77.o: in function `caml_data_segments':
:(.data+0x2468): undefined reference to `camlTest__data_begin'
ld: :(.data+0x2470): undefined reference to `camlTest__data_end'
ld: /tmp/camlstartupfa1c77.o: in function `caml_code_segments':
:(.data+0x2b30): undefined reference to `camlTest__code_begin'
ld: :(.data+0x2b38): undefined reference to `camlTest__code_end'
ld: /tmp/camlstartupfa1c77.o: in function `caml_frametable':
:(.data+0x2ec0): undefined reference to `camlTest__frametable'
ocamlopt: main.ml ---> main.cmx
ocamlopt -g -I lwt -I ubase -I system -I +unix -I +str -I system/generic -I lwt/generic -I /usr/pkg/lib/ocaml/site-lib/lablgtk3 -I /usr/pkg/lib/ocaml/site-lib/cairo2 -c /scratch/wip/unison-snapshot/work/unison-08dbd399b65b960e7f2ea356555209404d95024a/src/main.ml
ocamlopt: uigtk3.ml ---> uigtk3.cmx
ocamlopt -g -I lwt -I ubase -I system -I +unix -I +str -I system/generic -I lwt/generic -I /usr/pkg/lib/ocaml/site-lib/lablgtk3 -I /usr/pkg/lib/ocaml/site-lib/cairo2 -c /scratch/wip/unison-snapshot/work/unison-08dbd399b65b960e7f2ea356555209404d95024a/src/uigtk3.ml
File "caml_startup", line 1:
Error: Error during linking (exit code 1)
gmake[2]: *** [Makefile.OCaml:441: unison] Error 2
gmake[2]: Leaving directory '/scratch/wip/unison-snapshot/work/unison-08dbd399b65b960e7f2ea356555209404d95024a/src'
```

Three successive builds worked, then I saw:
```
ocamlopt -g -I lwt -I ubase -I system -I +unix -I +str -I system/generic -I lwt/generic -I /usr/pkg/lib/ocaml/site-lib
/lablgtk3 -I /usr/pkg/lib/ocaml/site-lib/cairo2 -c /scratch/wip/unison-snapshot/work/unison-08dbd399b65b960e7f2ea35655
5209404d95024a/src/uigtk3.ml                                                                                          
ld: /tmp/camlstartup12d7bb.o: in function `caml_startup__code_begin':                                                 
:(.text+0x665): undefined reference to `camlUitext__entry'                                                            
ld: /tmp/camlstartup12d7bb.o: in function `caml_globals':                                                             
:(.data+0x588): undefined reference to `camlUitext__gc_roots'                                                         
ld: /tmp/camlstartup12d7bb.o: in function `caml_data_segments':                                                       
:(.data+0x2458): undefined reference to `camlUitext__data_begin'                                                      
ld: :(.data+0x2460): undefined reference to `camlUitext__data_end'                                                    
ld: /tmp/camlstartup12d7bb.o: in function `caml_code_segments':                                                       
:(.data+0x2b20): undefined reference to `camlUitext__code_begin'                                                      
ld: :(.data+0x2b28): undefined reference to `camlUitext__code_end'                                                    
ld: /tmp/camlstartup12d7bb.o: in function `caml_frametable':                                                          
:(.data+0x2eb8): undefined reference to `camlUitext__frametable'                                                      
ld: linktext.o: in function `camlLinktext__code_begin':                                                               
:(.text+0x7): undefined reference to `camlUitext'                                                                     
ld: main.o: in function `camlMain__nonGuiStartup_840':                                                                
:(.text+0x9fc): undefined reference to `camlUitext__start_2798'                                                       
File "caml_startup", line 1:                                                                                          
Error: Error during linking (exit code 1)                                                                             
gmake[2]: *** [Makefile.OCaml:441: unison] Error 2
```

Five successful builds later, another error, this one looks different:
```
ocamlopt: system.ml ---> system.cmx                                                                                                                                                                                                                                                      
ocamlopt -g -I lwt -I ubase -I system -I +unix -I +str -I system/generic -I lwt/generic -c /scratch/wip/unison-snapshot/work/unison-08dbd399b65b960e7f2ea356555209404d95024a/src/system.ml
ocamlopt -g -I lwt -I ubase -I system -I +unix -I +str -I system/generic -I lwt/generic -I /usr/pkg/lib/ocaml/site-lib/lablgtk3 -I /usr/pkg/lib/ocaml/site-lib/cairo2 -c /scratch/wip/unison-snapshot/work/unison-08dbd399b65b960e7f2ea356555209404d95024a/src/ubase/util.mli
ocamlopt: tree.ml ---> tree.cmx                                       
ocamlopt -g -I lwt -I ubase -I system -I +unix -I +str -I system/generic -I lwt/generic -I /usr/pkg/lib/ocaml/site-lib/lablgtk3 -I /usr/pkg/lib/ocaml/site-lib/cairo2 -c /scratch/wip/unison-snapshot/work/unison-08dbd399b65b960e7f2ea356555209404d95024a/src/tree.ml    
File "/scratch/wip/unison-snapshot/work/unison-08dbd399b65b960e7f2ea356555209404d95024a/src/system.ml", line 1:                             
Error: Corrupted compilation unit description system/system_generic.cmx                                                                                                                                                                                                                  
ocamlopt: ubase/trace.mli ---> ubase/trace.cmi                        
ocamlopt -g -I lwt -I ubase -I system -I +unix -I +str -I system/generic -I lwt/generic -c /scratch/wip/unison-snapshot/work/unison-08dbd399b65b960e7f2ea356555209404d95024a/src/ubase/trace.mli                                                                       
gmake[2]: *** [Makefile.OCaml:433: system.cmx] Error 2                
gmake[2]: *** Waiting for unfinished jobs....                                                                                                                                                                                                                                            
ocamlopt: case.mli ---> case.cmi                                      
```
Full log:
[log2.txt](https://github.com/bcpierce00/unison/files/9304316/log2.txt)

Two later, 1a again:
```
ld: /tmp/camlstartup4b3500.o: in function `caml_startup__code_begin':                                              
:(.text+0x665): undefined reference to `camlUitext__entry'                                                            
ld: /tmp/camlstartup4b3500.o: in function `caml_globals':
:(.data+0x588): undefined reference to `camlUitext__gc_roots'
ld: /tmp/camlstartup4b3500.o: in function `caml_data_segments':
:(.data+0x2458): undefined reference to `camlUitext__data_begin'
ld: :(.data+0x2460): undefined reference to `camlUitext__data_end'
ld: /tmp/camlstartup4b3500.o: in function `caml_code_segments':
:(.data+0x2b20): undefined reference to `camlUitext__code_begin'
ld: :(.data+0x2b28): undefined reference to `camlUitext__code_end'
ld: /tmp/camlstartup4b3500.o: in function `caml_frametable':
:(.data+0x2eb8): undefined reference to `camlUitext__frametable'
ld: linktext.o: in function `camlLinktext__code_begin':
:(.text+0x7): undefined reference to `camlUitext'
ld: main.o: in function `camlMain__nonGuiStartup_840':
:(.text+0x9fc): undefined reference to `camlUitext__start_2798'
```

--- a/Makefile
+++ b/Makefile
@@ -17,7 +17,9 @@ docs:
 	$(MAKE) -C doc
 	$(MAKE) -C man
 
-manpage:
+# "src" is a prerequisite to prevent parallel build errors.
+# manpage builds currently require a pre-built "unison" binary.
+manpage: src
 	$(MAKE) -C man
 
 test:

